airflow/jobs.py

```diff
@@ -430,6 +465,29 @@ def process_dag(self, dag, executor):
                     if ti.is_queueable(flag_upstream_failed=True):
                         logging.debug('Queuing next run: ' + str(ti))
                         executor.queue_task_instance(ti, pickle_id=pickle_id)
+
+        # Checking state of active DagRuns
+        active_runs = session.query(models.DagRun).filter(
+            models.DagRun.dag_id == dag.dag_id,
+            models.DagRun.state == 'active'
+        ).all()
+        for run in active_runs:
+            logging.info("Checking state for " + str(run))
+            task_instances = session.query(TI).filter(
+                TI.dag_id == run.dag_id,
+                TI.execution_date == run.execution_date
+            ).all()
+            if len(task_instances) == len(dag.tasks):
```

Comment:
Can a dag have branches with tasks that are never executed within one run? If yes, then this check would not be sufficient.